### PR TITLE
Introduce an abstraction for handling stateful traversal of the tree

### PR DESCRIPTION
**Pull request by vil1**
While still being able to accumulate errors, produce partial results, etc.

The base principle is to build a *representation* of the computation (parsing, visiting, optimizing, generating, etc) that will eventually get `run`.  This representation keeps track of a state (of arbitrary type if we wanted to, but we practically stick to `RemorphContext` for now) along with the result it is producing.

This way, we can interleave any step of the computation with updates to this state, like counting the statements processed thus far (for any definition of what a statement is, including arbitrarily nested subqueries), the branch of the IR tree currently being processed, etc. 

Similarly, at any point of the computation, we can inspect the current state, to guide said computation (by customizing error messages with contextual information for example).

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/RemorphContext.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.databricks.labs.remorph
+
+import com.databricks.labs.remorph.intermediate.{LogicalPlan, TreeNode}
+import org.antlr.v4.runtime.ParserRuleContext
+
+sealed trait RemorphContext
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/RemorphContext.scala:6]:
```suggestion
sealed trait Conversion
```
what about this?

File: core/src/main/scala/com/databricks/labs/remorph/RemorphContext.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+case object Empty extends RemorphContext
+
+case class Sources(input: String) extends RemorphContext
+
+case class Parsed(tree: ParserRuleContext, sources: Option[Sources] = None) extends RemorphContext
```
[the remark this answers is not in the document]
Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/RemorphContext.scala:12]: In some case, like in tests, we want to "start" the process at, say, optimizer, so we may not have any `Sources` to provide

File: core/src/main/scala/com/databricks/labs/remorph/Result.scala [hunk 3]
```diff
@@ -10,6 +10,25 @@ object WorkflowStage {
   case object GENERATE extends WorkflowStage
 }
 
+final class TBA[S, +A](val runF: Result[S => Result[(S, A)]]) {
```
[the remark this answers is not in the document]
Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/Result.scala:13]: hmm that's definitely better than `TBA` but slightly misleading as the State isn't really the input of the computation (even though it is the input of the function here).

what do you think of 
`final class Stateful[State, Out](val computation: Result[State => Result[(State, Out)]])`
or
`final class Contextual[Context, Out](val computation: Result[Context => Result[(Context, Out)]])` ?

File: core/src/main/scala/com/databricks/labs/remorph/generators/package.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
+package com.databricks.labs.remorph
+
+import com.databricks.labs.remorph.intermediate.UncaughtException
+
+import scala.util.control.NonFatal
+
+package object generators {
+
+  implicit class TBAInterpolator(sc: StringContext) extends TBAS[RemorphContext] {
+    def tba(args: Any*): TBA[RemorphContext, String] = {
```
[the remark this answers is not in the document]
Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/generators/package.scala:10]: YES!

File: core/src/main/scala/com/databricks/labs/remorph/generators/package.scala [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  implicit class TBASeqOps(tbas: Seq[TBA[RemorphContext, String]]) extends TBAS[RemorphContext] {
+
+    def mkTba: TBA[RemorphContext, String] = mkTba("", "", "")
+
+    def mkTba(sep: String): TBA[RemorphContext, String] = mkTba("", sep, "")
+
+    def mkTba(start: String, sep: String, end: String): TBA[RemorphContext, String] = {
+      tbas.sequence.map(_.mkString(start, sep, end))
+    }
```
[the remark this answers is not in the document]
Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/generators/package.scala:50]: We sure can, I initially went down that path to make the initial refactoring (the one that introduced `Result`) as small as possible.

File: core/src/main/scala/com/databricks/labs/remorph/RemorphContext.scala (hunk 1 above)
Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/RemorphContext.scala:6]: ..., software version 7.0 ...

I agree, it makes better sense

File: core/src/main/scala/com/databricks/labs/remorph/Result.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (nfx) [reply core/src/main/scala/com/databricks/labs/remorph/Result.scala:13]: oh, can we name it as `Transformation`

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**